### PR TITLE
Refactors emotes to use datums.

### PR DESCRIPTION
**Pull request by ShiftyRail**
Adding *dab has never been easier.

Largely ported from TG, with a few modifications (we have a different *scream, we have *fart, and a few exclusive silicon emotes.)

I also removed the following line :

```dm
msg = "<b>[H]</b> lets out a [pick("disgusting","revolting","horrible","strangled","god awful")] noise out of \his mutilated asshole."
```

Which happened when you tried to fart without an arse
Jesus Christ
I'll PR to remove it as soon as this gets merged

:cl:
- experimental: The system we use to recognise and display emotes has largely been refactored. Please report any emote missing or being displayed improperly.

### CODE REVIEW

**Comment by beesil** [comment]
*dab

File: code/modules/mob/living/silicon/emote.dm [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	key_third_person = "chimes"
+	message = "chimes."
+	sound = 'sound/machines/chime.ogg'
+
+/datum/emote/sound/silicon/honk
+	key = "honk"
+	key_third_person = "honks"
+	message = "honks."
+	vary = TRUE
+	sound = 'sound/items/bikehorn.ogg'
```
Comment (SonixApache) [inline code/modules/mob/living/silicon/emote.dm]: Shouldn't this only be available to clownborgs?

Comment (jknpj) [reply code/modules/mob/living/silicon/emote.dm]: check MODULE_IS_A_CLOWN here

File: code/modules/mob/living/silicon/emote.dm [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+/datum/emote/sound/silicon/chime
+	key = "chime"
+	key_third_person = "chimes"
+	message = "chimes."
+	sound = 'sound/machines/chime.ogg'
+
+/datum/emote/sound/silicon/sad
+	key = "sad"
+	message = "plays a sad trombone..."
+	sound = 'sound/misc/sadtrombone.ogg'
```
Comment (jknpj) [inline code/modules/mob/living/silicon/emote.dm]: maybe here too

**Comment by SirSkeleto** [comment]
*dab

File: code/modules/mob/living/silicon/emote.dm [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+/datum/emote/sound/silicon/law/can_run_emote(var/mob/user, var/status_check = TRUE)
+	. = ..()
+	var/mob/living/silicon/robot/R = user
+	if (!istype(R))
+		return FALSE
+	if (!(R.module && (R.module.quirk_flags & MODULE_IS_THE_LAW)))
+		return FALSE
+
+/datum/emote/sound/silicon/mommi
```
Comment (jknpj) [inline code/modules/mob/living/silicon/emote.dm]: move mommi emotes to their own folder

File: code/modules/mob/living/silicon/emote.dm [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+/datum/emote/sound/silicon/mommi/comment
+	key = "comment"
+	message = "vocalizes."
+	message_param = "cheerily vocalizes at %t."
+
+/datum/emote/sound/silicon/mommi/comment/run_emote(mob/user, params, type_override)
+	sound = get_sfx("mommicomment")
+
+/mob/living/silicon/robot/verb/powerwarn()
```
Comment (ShiftyRail) [inline code/modules/mob/living/silicon/emote.dm]: Unatomic

Comment (SonixApache) [reply code/modules/mob/living/silicon/emote.dm]: ported from TG, right?

File: code/modules/mob/living/silicon/emote.dm [hunk 5]
```diff
@@ -0,0 +1,100 @@
+/datum/emote/silicon
+	mob_type_allowed_typecache = list(/mob/living/silicon)
+	emote_type = EMOTE_AUDIBLE
+	var/module_required
+
+/datum/emote/sound/silicon
+	mob_type_allowed_typecache = list(/mob/living/silicon)
+	emote_type = EMOTE_AUDIBLE
+	var/module_required
```
Comment (jknpj) [inline code/modules/mob/living/silicon/emote.dm]: module_quirk_required is better, since we're checking for flags here

**Comment by cadaverouscoyote** [comment]
> I also removed the following line :
`msg = "<b>[H]</b> lets out a [pick("disgusting","revolting","horrible","strangled","god awful")] noise out of \his mutilated asshole."`

where the feature loss tag

**Comment by tolan1** [comment]
*dab

File: code/modules/mob/emote.dm [hunk 6]
```diff
@@ -1,56 +1,52 @@
...
-		else if (m_type & 2)
-			for(var/mob/M in get_hearers_in_view(7, src))
-				M.show_message(message, m_type)
+	E.run_emote(src, param, m_type)
+
+/datum/emote/flip
+	key = "flip"
+	key_third_person = "flips"
+	restraint_check = TRUE
+	mob_type_allowed_typecache = list(/mob/living, /mob/dead/observer)
```
Comment (DamianX) [inline code/modules/mob/emote.dm]: we don't have typecaches though

Comment (ShiftyRail) [reply code/modules/mob/emote.dm]: TG nerds have it
I'll do a mass search and replace if it's needed

**Comment by strawberrytan** [comment]
Feature loss of inability to spam chat window with mutilated ass

File: code/modules/mob/emote.dm (hunk 6 above)
Comment (DamianX) [reply code/modules/mob/emote.dm]: I don't care, it is a factually wrong name though

**Comment by Exxion** [comment]
Can't say I care for that line but it's still unatomic
Make a separate PR to remove it

**Comment by ShiftyRail** [comment]
Goddamit

File: code/modules/mob/living/emote.dm [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+	message_robot = "shudders violently for a moment before falling still, its eyes slowly darkening."
+	message_AI = "lets out a flurry of sparks, its screen flickering as its systems slowly halt."
+	message_alien = "lets out a waning guttural screech, green blood bubbling from its maw..."
+	message_larva = "lets out a sickly hiss of air and falls limply to the floor..."
+	message_monkey = "lets out a faint chimper as it collapses and stops moving..."
+	message_simple =  "stops moving..."
+	stat_allowed = UNCONSCIOUS
+
+/datum/emote/living/deathgasp/run_emote(mob/user, params)
+	if(. && isalienadult(user))
```
Comment (DamianX) [inline code/modules/mob/living/emote.dm]: `if(null & ...)` ???

File: code/modules/mob/living/silicon/mommi/emote.dm [hunk 8]
```diff
@@ -1,247 +1,11 @@
...
-		if (m_type & 1)
-			for(var/mob/O in viewers(src, null))
-				O.show_message(message, m_type)
-		else
-			for(var/mob/O in hearers(src, null))
-				O.show_message(message, m_type)
-	return
\ No newline at end of file
+/datum/emote/sound/silicon/mommi/comment/run_emote(mob/user, params, type_override)
+	sound = get_sfx("mommicomment")
```
Comment (DamianX) [inline code/modules/mob/living/silicon/mommi/emote.dm]: don't you have to call the parent?
Or even just assign this in New and pick from the list in a more generic `run_emote`

**Review by DamianX — DISMISSED** [review]

File: code/modules/mob/living/silicon/mommi/emote.dm (hunk 8 above)
Comment (ShiftyRail) [reply code/modules/mob/living/silicon/mommi/emote.dm]: I think both options are equivalent and this is the cleanest one.
Although I did need to call the parent here

**Comment by ShiftyRail** [comment]
Did the requested changes and made the PR atomic

File: code/modules/mob/living/silicon/emote.dm (hunk 5 above)
Comment (jknpj) [reply code/modules/mob/living/silicon/emote.dm]: FIXIT

**Pull request merged**